### PR TITLE
Implement force-inline optimizer option

### PR DESCRIPTION
**Pull request by c0d1f1ed**
SwiftShader does not support function calls in shaders, and has relied
on the spirvtools::opt::InlineExhaustivePass to eliminate them through
inlining. But since #3858, functions with the DontInline control bit do
not get inlined. This makes SwiftShader fail some of the dEQP-VK.
spirv_assembly. instruction.compute.function_control.* tests

This change adds an option to the optimizer to force-inline functions
whenever possible, effectively ignoring DontInline. Note this still
produces spec-compliant SPIR-V code, since DontInline is a "Strong
request, to the extent possible, to not inline the function." but not a
demand.

Bug: https://issuetracker.google.com/issues/172212963

### CODE REVIEW

**Review by alan-baker — APPROVED** [review]

**Comment by s-perron** [comment]
I'm wondering if it would be better to implement a pass that removes the Don't Inline tag.  This makes each pass conceptually simpler because you won't see the inlining pass inline a function marked as Don't Inline.

**Review by s-perron — CHANGES_REQUESTED** [review]
I'd rather see a pass that modifies the OpLoopMerge instead of an option that modifies the behaviour of the inlining passes.

**Comment by c0d1f1ed** [comment]
> I'm wondering if it would be better to implement a pass that removes the Don't Inline tag. This makes each pass conceptually simpler because you won't see the inlining pass inline a function marked as Don't Inline.

Right, that might avoid some confusion. As a counter-example though, the C++ `inline` keyword makes no guarantees at all about functions actually getting inlined, nor does its omission prevent inlining. Optimization flags have a much bigger effect on inlining behavior. So I'd argue that people (app developers as well as compiler developers) are well accustomed to seeing attributes like this getting overridden<sup>1</sup>.

I'm not sure I agree with keeping things simpler if it was implemented as a separate pass. This is effectively a 2-line change. The rest is just plumbing. A separate pass is going to be a larger change, and make things slower by iterating over the entire binary and rewriting things. It would have a low signal-to-noise ratio. My current change is very concise and focused on the issue at hand. Note also that it can still be rewritten as a separate pass, in case the complexity increases and the pros start to outweigh the cons, but that seems relatively unlikely to me. I'm happy to take responsibility over maintaining this functionality.

1. Granted, C++ explicitly redefined `inline` to be about preventing ODR violation, instead of inlining of function calls, while SPIR-V defines it as a strong hint to perform the latter, but I think my argument about things ultimately being up to the optimization pass still stands.

**Comment by s-perron** [comment]
The problem is being able to control inlining the way one would like.  This is a global option that will apply to every invocation of inlining.  You may be okay with that, but others may not.  For example, someone may write a pass that will create a new function that is not supposed to be inlined for whatever reason.  They may want to inline all others.  If it was a standalone pass, then that pass could be run first, then create their new functions, and then run inlining.  That flexibly does not exist in this implementation.  

Another complication with this implementation is that somebody might decide to modify the global option in some other pass.  If this causes a problem, it will be annoying to track down.  I'd like to minimize the number of ways someone can create a problem for themselves.  

At the very least, it has to be an option for the transformation, not an option stored in the context.

Finally, the context is suppose to be a wrapper around the module that adds commonly used functions around it.  I know this is broken because of the `preserve_bindings_` and `preserve_spec_constants_`, which seems to be what you are following.  I would like to have them done differently.  Those are harder because they have to apply to every invocation of ADCE, or something like '-O' become unusable if you need a suboption to ADCE.  Another difference is that they stop the optimizer from doing something.  In your case, you are asking the optimizer to do something: ignore "Don't Inline".  This is why a separate pass seems to make more sense.

To sum up, I believe this should be a pass because it will 1) offer more flexibility, 2) be easier to debug, and 3) avoid stored even more state in the ir context that cannot be derived from the module.

**Comment by c0d1f1ed** [comment]
> The problem is being able to control inlining the way one would like. This is a global option that will apply to every invocation of inlining. You may be okay with that, but others may not. For example, someone may write a pass that will create a new function that is not supposed to be inlined for whatever reason. They may want to inline all others. If it was a standalone pass, then that pass could be run first, then create their new functions, and then run inlining. That flexibly does not exist in this implementation.

Thanks for explaining your rationale. It appears to be based on hypotheticals though. I don't think that should guide our design, as this is the primary cause of feature creep and code bloat. Note that prior to my PR, nobody (to my knowledge) was asking for this kind of scenario you describe, to be supported. So it shouldn't suddenly be a design requirement (also I think it would be a bad requirement, but more on that later).

In contrast, I have an immediate need for this PR as-is, to make a Vulkan 1.2 conformance submission with Khronos for SwiftShader. Just to be clear, I don't mean this relative urgency to justify bad design, if it can be shown to be harmful to today's or *known* future needs. I'll put in the extra work if necessary to avoid adding technical debt. But I also don't think we should delay things and add bloat based on hypotheticals.

> Another complication with this implementation is that somebody might decide to modify the global option in some other pass. If this causes a problem, it will be annoying to track down. I'd like to minimize the number of ways someone can create a problem for themselves.

I definitely support avoiding any footguns, but I don't think this scenario is realistic, nor do I think it would lead to issues that are hard to debug, and it should have a different solution. Note that for the foreseeable future, SwiftShader will be the only user of this force-inline feature. It serves one little purpose. So it's unlikely to undergo significant modification by others that would cause them to run into an issue that is difficult to debug. It seems to me that any and all issues caused by modifying global options, can be fixed by making them immutable within passes. That seems like a footgun avoidance measure worth spending time on, but is out-of-scope for this PR since there is "prior art".

> At the very least, it has to be an option for the transformation, not an option stored in the context.

I disagree. The option needs to be global because it's a flag that says to the optimizer, as a whole, to force-inline all function calls, because our Vulkan implementation can't handle `OpFunctionCall`. This design requirement also means an implementation based on a separate pass would be flawed, since there's no guarantee that after we've removed `DontInline`, we'll actually see all functions getting inlined.

**Comment by s-perron** [comment]
To have everything inlined, you only need one pass of inlining that will inline everything.  The `InlineExhaustivePass` is intended to do that.  It will inline everything that it is allowed to inline.  The DontInline function control is sort of an exception.  We could add an option to that pass that will ignore the DontInline function control.  This can be passed down to InlinePass.

Also make sure that you update the "name" of the pass as is done in scalar replacement so that the name of the pass is the command line option needed to rerun the pass the same way.

Then on the swiftshader side you can run 4 passes:

```
      // Wrap OpKill instructions so all other code can be inlined.
      RegisterPass(CreateWrapOpKillPass())
          // Remove unreachable block so that merge return works.
          .RegisterPass(CreateDeadBranchElimPass())
          // Merge the returns so we can inline.
          .RegisterPass(CreateMergeReturnPass())
          // Make sure uses and definitions are in the same function.
          .RegisterPass(CreateInlineExhaustivePass(/*ignore_dont_inline =*/ true));
```

This will inline all function that can be legally inlined.  It will sometimes leaves calls to a function that contains only an OpKill, which I believe SwiftShader has special cased.  You can register any other passes you want after that as well, include the predefined sets.

I should be a smaller change than adding a new pass.  It will accomplish what you want.  It also fits better with the design in spirv-opt.  I hope you find this compromise acceptable.

**Comment by dneto0** [comment]
>Thanks for explaining your rationale. It appears to be based on hypotheticals though.

It's not quite a hypothetical.  In the normal course of transformations, we wrap OpKill in a function, and then call that. This is needed for the case where the OpKill might appear in a continue construct: recall that the backedge block must post-dominate the continue target, and OpKill in there breaks that rule.


> In contrast, I have an immediate need for this PR as-is, to make a Vulkan 1.2 conformance submission with Khronos for SwiftShader. 

So take Steven's advice.

>  Note that for the foreseeable future, SwiftShader will be the only user of this force-inline feature. It serves one little purpose. 

It's a design smell to rely on assumptions about the context of use.
Steven's original suggestion (make a pass to strip DontInline) is the simplest, most composable design.
